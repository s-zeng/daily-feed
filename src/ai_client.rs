use serde::{Deserialize, Serialize};
use std::error::Error;
use std::fmt;

#[derive(Debug)]
pub enum AiClientError {
    RequestError(String),
    ParseError(String),
    ConfigError(String),
}

impl fmt::Display for AiClientError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            AiClientError::RequestError(msg) => write!(f, "Request error: {}", msg),
            AiClientError::ParseError(msg) => write!(f, "Parse error: {}", msg),
            AiClientError::ConfigError(msg) => write!(f, "Config error: {}", msg),
        }
    }
}

impl Error for AiClientError {}

#[derive(Debug, Clone)]
pub enum AiProvider {
    Ollama { base_url: String, model: String },
    Anthropic { api_key: String, model: String },
}

#[derive(Serialize)]
struct ChatCompletionRequest {
    model: String,
    messages: Vec<ChatMessage>,
    temperature: f32,
    stream: bool,
}

#[derive(Serialize, Deserialize)]
struct ChatMessage {
    role: String,
    content: String,
}

#[derive(Deserialize)]
struct ChatCompletionResponse {
    choices: Vec<ChatChoice>,
}

#[derive(Deserialize)]
struct ChatChoice {
    message: ChatMessage,
}

pub struct AiClient {
    provider: AiProvider,
    client: reqwest::Client,
}

impl AiClient {
    pub fn new(provider: AiProvider) -> Result<Self, AiClientError> {
        let client = reqwest::Client::new();
        Ok(AiClient { provider, client })
    }

    pub async fn generate_text(&self, prompt: &str) -> Result<String, AiClientError> {
        match &self.provider {
            AiProvider::Ollama { base_url, model } => {
                self.call_ollama(base_url, model, prompt).await
            }
            AiProvider::Anthropic { api_key, model } => {
                self.call_anthropic(api_key, model, prompt).await
            }
        }
    }

    async fn call_ollama(&self, base_url: &str, model: &str, prompt: &str) -> Result<String, AiClientError> {
        let url = format!("{}/v1/chat/completions", base_url);
        let request = ChatCompletionRequest {
            model: model.to_string(),
            messages: vec![ChatMessage {
                role: "user".to_string(),
                content: prompt.to_string(),
            }],
            temperature: 0.0,
            stream: false,
        };

        let response = self
            .client
            .post(&url)
            .json(&request)
            .send()
            .await
            .map_err(|e| AiClientError::RequestError(e.to_string()))?;

        if !response.status().is_success() {
            return Err(AiClientError::RequestError(format!(
                "HTTP {}: {}",
                response.status(),
                response.text().await.unwrap_or_default()
            )));
        }

        let chat_response: ChatCompletionResponse = response
            .json()
            .await
            .map_err(|e| AiClientError::ParseError(e.to_string()))?;

        match chat_response.choices.first() {
            Some(choice) => Ok(choice.message.content.clone()),
            None => Err(AiClientError::ParseError("No choices in response".to_string())),
        }
    }

    async fn call_anthropic(&self, _api_key: &str, _model: &str, _prompt: &str) -> Result<String, AiClientError> {
        // TODO: Implement Anthropic API integration
        // For now, return an error indicating it's not implemented
        Err(AiClientError::ConfigError("Anthropic API integration not yet implemented".to_string()))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_ollama_client_creation() {
        let provider = AiProvider::Ollama {
            base_url: "http://127.0.0.1:1234".to_string(),
            model: "llama2".to_string(),
        };
        
        let client = AiClient::new(provider);
        assert!(client.is_ok());
    }

    #[tokio::test]
    async fn test_anthropic_client_creation() {
        let provider = AiProvider::Anthropic {
            api_key: "test-key".to_string(),
            model: "claude-3-sonnet-20240229".to_string(),
        };
        
        let client = AiClient::new(provider);
        assert!(client.is_ok());
    }
}