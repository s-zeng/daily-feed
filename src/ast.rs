use nonempty::NonEmpty;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ArticleContent {
    pub blocks: NonEmpty<ContentBlock>,
    pub reading_time_minutes: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct FeedContent {
    pub articles: NonEmpty<Article>,
    pub total_reading_time_minutes: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct DocumentContent {
    pub feeds: NonEmpty<Feed>,
    pub total_reading_time_minutes: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Headline {
    pub title: String,
    pub published_date: Option<String>,
    pub source_name: String,
    pub url: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Document {
    pub metadata: DocumentMetadata,
    pub front_page: Option<Vec<ContentBlock>>,
    pub content: Option<DocumentContent>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct DocumentMetadata {
    pub title: String,
    pub author: String,
    pub description: Option<String>,
    pub generated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Feed {
    pub name: String,
    pub description: Option<String>,
    pub url: Option<String>,
    pub content: Option<FeedContent>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Article {
    pub title: String,
    pub metadata: ArticleMetadata,
    pub comments: Vec<Comment>,
    pub content: Option<ArticleContent>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ArticleMetadata {
    pub published_date: Option<String>,
    pub author: Option<String>,
    pub url: Option<String>,
    pub feed_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Comment {
    pub author: String,
    pub content: Vec<ContentBlock>,
    pub upvotes: u32,
    pub downvotes: u32,
    pub timestamp: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum ContentBlock {
    Paragraph(TextContent),
    Heading {
        level: u8,
        content: TextContent,
    },
    List {
        ordered: bool,
        items: Vec<TextContent>,
    },
    Quote(TextContent),
    Code {
        language: Option<String>,
        content: String,
    },
    Link {
        url: String,
        text: String,
    },
    Image {
        url: String,
        alt: Option<String>,
    },
    Raw(String), // For complex HTML that we want to preserve as-is
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct TextContent {
    pub spans: Vec<TextSpan>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct TextSpan {
    pub text: String,
    pub formatting: TextFormatting,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Default)]
pub struct TextFormatting {
    pub bold: bool,
    pub italic: bool,
    pub code: bool,
    pub link: Option<String>,
}

impl Document {
    pub fn new(title: String, author: String) -> Self {
        Self {
            metadata: DocumentMetadata {
                title,
                author,
                description: None,
                generated_at: chrono::Utc::now().to_rfc3339(),
            },
            front_page: None,
            content: None,
        }
    }

    pub fn set_content(&mut self, feeds: NonEmpty<Feed>, total_reading_time_minutes: u32) {
        self.content = Some(DocumentContent {
            feeds,
            total_reading_time_minutes,
        });
    }

    pub fn set_front_page(&mut self, front_page: Vec<ContentBlock>) {
        self.front_page = Some(front_page);
    }

    pub fn total_articles(&self) -> usize {
        self.content
            .as_ref()
            .map(|c| c.feeds.iter().map(|f| f.content.as_ref().map_or(0, |fc| fc.articles.len())).sum())
            .unwrap_or(0)
    }

    pub fn extract_headlines(&self) -> Vec<Headline> {
        self.content
            .as_ref()
            .map(|c| {
                c.feeds
                    .iter()
                    .flat_map(|feed| {
                        feed.content.as_ref().map(|fc| {
                            fc.articles.iter().map(|article| Headline {
                                title: article.title.clone(),
                                published_date: article.metadata.published_date.clone(),
                                source_name: article.metadata.feed_name.clone(),
                                url: article.metadata.url.clone(),
                            }).collect::<Vec<_>>()
                        }).unwrap_or_default()
                    })
                    .collect()
            })
            .unwrap_or_default()
    }
}

impl Feed {
    pub fn new(name: String) -> Self {
        Self {
            name,
            description: None,
            url: None,
            content: None,
        }
    }

    pub fn with_description(mut self, description: String) -> Self {
        self.description = Some(description);
        self
    }

    pub fn with_url(mut self, url: String) -> Self {
        self.url = Some(url);
        self
    }

    pub fn set_content(&mut self, articles: NonEmpty<Article>, total_reading_time_minutes: u32) {
        self.content = Some(FeedContent {
            articles,
            total_reading_time_minutes,
        });
    }
}

impl Article {
    pub fn new(title: String, feed_name: String) -> Self {
        Self {
            title,
            metadata: ArticleMetadata {
                published_date: None,
                author: None,
                url: None,
                feed_name,
            },
            comments: Vec::new(),
            content: None,
        }
    }

    pub fn set_content(&mut self, blocks: NonEmpty<ContentBlock>, reading_time_minutes: u32) {
        self.content = Some(ArticleContent {
            blocks,
            reading_time_minutes,
        });
    }

    pub fn with_published_date(mut self, date: String) -> Self {
        self.metadata.published_date = Some(date);
        self
    }

    pub fn with_url(mut self, url: String) -> Self {
        self.metadata.url = Some(url);
        self
    }

    pub fn add_comment(&mut self, comment: Comment) {
        self.comments.push(comment);
    }
}

impl TextContent {
    pub fn plain(text: String) -> Self {
        Self {
            spans: vec![TextSpan {
                text,
                formatting: TextFormatting::default(),
            }],
        }
    }

    pub fn from_spans(spans: Vec<TextSpan>) -> Self {
        Self { spans }
    }

    pub fn is_empty(&self) -> bool {
        self.spans.iter().all(|span| span.text.trim().is_empty())
    }

    pub fn to_plain_text(&self) -> String {
        self.spans.iter().map(|span| span.text.as_str()).collect()
    }
}

impl TextSpan {
    pub fn plain(text: String) -> Self {
        Self {
            text,
            formatting: TextFormatting::default(),
        }
    }

    pub fn bold(text: String) -> Self {
        Self {
            text,
            formatting: TextFormatting {
                bold: true,
                ..Default::default()
            },
        }
    }

    pub fn italic(text: String) -> Self {
        Self {
            text,
            formatting: TextFormatting {
                italic: true,
                ..Default::default()
            },
        }
    }

    pub fn code(text: String) -> Self {
        Self {
            text,
            formatting: TextFormatting {
                code: true,
                ..Default::default()
            },
        }
    }

    pub fn link(text: String, url: String) -> Self {
        Self {
            text,
            formatting: TextFormatting {
                link: Some(url),
                ..Default::default()
            },
        }
    }
}
