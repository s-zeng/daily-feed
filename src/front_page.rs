use crate::ai_client::{AiClient, AiClientError, AiProvider};
use crate::ast::{ContentBlock, Document, TextContent, TextFormatting, TextSpan};
use serde::{Deserialize, Serialize};
use std::error::Error;
use std::fmt;

#[derive(Debug)]
pub enum FrontPageError {
    AiError(AiClientError),
    GenerationError(String),
    ParseError(String),
}

impl fmt::Display for FrontPageError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            FrontPageError::AiError(e) => write!(f, "AI error: {}", e),
            FrontPageError::GenerationError(msg) => write!(f, "Generation error: {}", msg),
            FrontPageError::ParseError(msg) => write!(f, "Parse error: {}", msg),
        }
    }
}

impl Error for FrontPageError {}

impl From<AiClientError> for FrontPageError {
    fn from(error: AiClientError) -> Self {
        FrontPageError::AiError(error)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StructuredFrontPage {
    pub theme: String,
    pub sources: Vec<SourceSummary>,
    pub context: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SourceSummary {
    pub name: String,
    pub summary: String,
    pub key_stories: Vec<String>,
}

pub struct FrontPageGenerator {
    ai_client: AiClient,
}

impl FrontPageGenerator {
    pub fn new(provider: AiProvider) -> Result<Self, FrontPageError> {
        let ai_client = AiClient::new(provider)?;
        Ok(FrontPageGenerator { ai_client })
    }

    pub async fn generate_structured_front_page_from_document(
        &self,
        document: &Document,
    ) -> Result<Vec<ContentBlock>, FrontPageError> {
        let structured_data = self.generate_structured_data_by_source(document).await?;
        Ok(self.convert_to_ast(&structured_data))
    }

    async fn generate_structured_data_by_source(
        &self,
        document: &Document,
    ) -> Result<StructuredFrontPage, FrontPageError> {
        let content = self.prepare_content_by_source(document)?;
        let prompt = self.build_structured_prompt_by_source(&content);

        let response = self.ai_client.generate_text(&prompt).await?;
        self.parse_structured_response_by_source(&response)
    }

    pub fn convert_to_ast(&self, front_page: &StructuredFrontPage) -> Vec<ContentBlock> {
        let mut blocks = Vec::new();

        // Add theme as opening paragraph
        blocks.push(ContentBlock::Paragraph(TextContent::from_spans(vec![
            TextSpan {
                text: "Today's World: ".to_string(),
                formatting: TextFormatting {
                    bold: true,
                    ..Default::default()
                },
            },
            TextSpan::plain(front_page.theme.clone()),
        ])));

        // Add each source summary
        for source in &front_page.sources {
            // Add source heading
            blocks.push(ContentBlock::Heading {
                level: 2,
                content: TextContent::plain(source.name.clone()),
            });

            // Add source summary
            blocks.push(ContentBlock::Paragraph(TextContent::plain(
                source.summary.clone(),
            )));

            // Add key stories if present
            if !source.key_stories.is_empty() {
                blocks.push(ContentBlock::Heading {
                    level: 3,
                    content: TextContent::plain("Key Stories".to_string()),
                });

                let story_items: Vec<TextContent> = source
                    .key_stories
                    .iter()
                    .map(|story| TextContent::plain(story.clone()))
                    .collect();

                blocks.push(ContentBlock::List {
                    ordered: false,
                    items: story_items,
                });
            }
        }

        // Add context if present
        if let Some(context) = &front_page.context {
            blocks.push(ContentBlock::Heading {
                level: 2,
                content: TextContent::plain("Looking Ahead".to_string()),
            });
            blocks.push(ContentBlock::Paragraph(TextContent::plain(context.clone())));
        }

        blocks
    }

    pub fn extract_json_from_response(&self, response: &str) -> String {
        // Look for JSON in markdown code blocks first
        if let Some(start) = response.find("```json") {
            let after_start = &response[start + 7..]; // Skip "```json"
            if let Some(end) = after_start.find("```") {
                return after_start[..end].trim().to_string();
            }
        }

        // Look for JSON in generic code blocks
        if let Some(start) = response.find("```") {
            let after_start = &response[start + 3..];
            if let Some(end) = after_start.find("```") {
                let content = after_start[..end].trim();
                // Check if this looks like JSON (starts with { and ends with })
                if content.trim_start().starts_with('{') && content.trim_end().ends_with('}') {
                    return content.to_string();
                }
            }
        }

        // Look for standalone JSON objects (lines starting with { and ending with })
        let lines: Vec<&str> = response.lines().collect();
        let mut json_start = None;
        let mut json_end = None;
        let mut brace_count = 0;

        for (i, line) in lines.iter().enumerate() {
            let trimmed = line.trim();
            if json_start.is_none() && trimmed.starts_with('{') {
                json_start = Some(i);
                brace_count = 1;
                // Count braces in the first line
                for ch in trimmed.chars().skip(1) {
                    match ch {
                        '{' => brace_count += 1,
                        '}' => brace_count -= 1,
                        _ => {}
                    }
                }
                if brace_count == 0 {
                    json_end = Some(i);
                    break;
                }
            } else if json_start.is_some() && brace_count > 0 {
                // Count braces in subsequent lines
                for ch in trimmed.chars() {
                    match ch {
                        '{' => brace_count += 1,
                        '}' => brace_count -= 1,
                        _ => {}
                    }
                }
                if brace_count == 0 {
                    json_end = Some(i);
                    break;
                }
            }
        }

        if let (Some(start), Some(end)) = (json_start, json_end) {
            return lines[start..=end].join("\n");
        }

        // If no JSON found, return original response
        response.to_string()
    }

    pub fn parse_structured_response_by_source(
        &self,
        response: &str,
    ) -> Result<StructuredFrontPage, FrontPageError> {
        // First, try to extract JSON from markdown code blocks
        let json_content = self.extract_json_from_response(response);

        // Try to parse as JSON first
        if let Ok(structured) = serde_json::from_str::<StructuredFrontPage>(&json_content) {
            return Ok(structured);
        }

        // If JSON parsing fails, try to extract structured data from markdown-like format
        self.parse_markdown_response_by_source(response)
    }

    fn parse_markdown_response_by_source(
        &self,
        response: &str,
    ) -> Result<StructuredFrontPage, FrontPageError> {
        let lines: Vec<&str> = response.lines().collect();
        let mut theme = String::new();
        let mut sources = Vec::new();
        let mut context = None;

        let mut current_section = "theme";
        let mut current_source: Option<SourceSummary> = None;

        for line in lines {
            let line = line.trim();

            if line.is_empty() {
                continue;
            }

            // Detect section headers
            if line.contains("Today's World") || line.contains("**Today's World**") {
                current_section = "theme";
                // Extract theme from same line if present
                let clean_line = line
                    .replace("**Today's World**:", "")
                    .replace("**Today's World**: ", "")
                    .replace("Today's World:", "")
                    .replace("Today's World: ", "")
                    .trim()
                    .to_string();
                if !clean_line.is_empty() {
                    theme = clean_line;
                }
                continue;
            } else if line.starts_with("##")
                || line.contains("**") && !line.contains("Looking Ahead")
            {
                // This might be a source name
                if let Some(source) = current_source.take() {
                    sources.push(source);
                }

                let source_name = line
                    .replace("##", "")
                    .replace("**", "")
                    .replace(":", "")
                    .trim()
                    .to_string();

                current_source = Some(SourceSummary {
                    name: source_name,
                    summary: String::new(),
                    key_stories: Vec::new(),
                });
                current_section = "source";
                continue;
            } else if line.contains("Looking Ahead") || line.contains("**Looking Ahead**") {
                if let Some(source) = current_source.take() {
                    sources.push(source);
                }
                current_section = "context";
                continue;
            }

            match current_section {
                "theme" => {
                    if !theme.is_empty() {
                        theme.push(' ');
                    }
                    let clean_line = line
                        .replace("**Today's World**:", "")
                        .replace("**Today's World**: ", "")
                        .replace("Today's World:", "")
                        .replace("Today's World: ", "")
                        .trim()
                        .to_string();
                    if !clean_line.is_empty() {
                        theme.push_str(&clean_line);
                    }
                }
                "source" => {
                    if let Some(ref mut source) = current_source {
                        if line.starts_with("• ")
                            || line.starts_with("- ")
                            || line.starts_with("* ")
                        {
                            // This is a key story
                            let story = line
                                .trim_start_matches(&['•', '-', '*', ' '][..])
                                .trim()
                                .to_string();
                            source.key_stories.push(story);
                        } else {
                            // This is part of the summary
                            if !source.summary.is_empty() {
                                source.summary.push(' ');
                            }
                            source.summary.push_str(line);
                        }
                    }
                }
                "context" => {
                    if context.is_none() {
                        context = Some(String::new());
                    }
                    if let Some(ref mut ctx) = context {
                        if !ctx.is_empty() {
                            ctx.push(' ');
                        }
                        ctx.push_str(line);
                    }
                }
                _ => {}
            }
        }

        // Save the last source
        if let Some(source) = current_source {
            sources.push(source);
        }

        if theme.is_empty() && sources.is_empty() {
            return Err(FrontPageError::ParseError(
                "Could not parse structured front page from AI response".to_string(),
            ));
        }

        Ok(StructuredFrontPage {
            theme: if theme.is_empty() {
                "Multiple developing stories shape today's landscape".to_string()
            } else {
                theme
            },
            sources,
            context,
        })
    }

    pub fn build_structured_prompt_by_source(&self, content: &str) -> String {
        format!(
            r#"You are a senior news editor creating a structured "Front Page" summary organized by news sources. 

Analyze the provided content and return a JSON response with this exact structure:

{{
  "theme": "One sentence capturing the day's most significant theme or development across all sources",
  "sources": [
    {{
      "name": "Source name",
      "summary": "2-3 sentences summarizing the main themes and developments from this source",
      "key_stories": ["Key story title 1", "Key story title 2", "Key story title 3"]
    }}
  ],
  "context": "Optional sentence connecting stories across sources to broader trends"
}}

Guidelines:
- For each source, provide a thematic summary of their coverage
- Include 2-4 most important story titles from each source
- Maintain neutral tone
- Focus on what each source is emphasizing or covering uniquely
- Keep source summaries concise but informative
- The overall theme should reflect patterns across all sources

Daily feed content organized by source:
{}

Return only valid JSON with the structure above."#,
            content
        )
    }

    pub fn prepare_content_by_source(&self, document: &Document) -> Result<String, FrontPageError> {
        let mut content = String::new();

        for feed in &document.feeds {
            content.push_str(&format!("# Source: {}\n", feed.name));

            if let Some(description) = &feed.description {
                content.push_str(&format!("**Description:** {}\n", description));
            }

            if let Some(url) = &feed.url {
                content.push_str(&format!("**URL:** {}\n", url));
            }

            content.push_str("\n**Articles:**\n");

            for article in &feed.articles {
                content.push_str(&format!("- {}", article.title));

                if let Some(date) = &article.metadata.published_date {
                    content.push_str(&format!(" ({})", date));
                }

                content.push_str("\n");
            }

            content.push_str("\n");
        }

        Ok(content)
    }
}
