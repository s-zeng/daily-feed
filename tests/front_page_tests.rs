use daily_feed::ai_client::AiProvider;
use daily_feed::ast::{
    Article, ArticleMetadata, ContentBlock, Document, DocumentMetadata, Feed, TextContent,
};
use daily_feed::front_page::FrontPageGenerator;
use insta::assert_snapshot;

fn create_test_document() -> Document {
    let breaking_article = Article {
        title: "Major Tech Company Announces Revolutionary AI Breakthrough".to_string(),
        content: vec![
            ContentBlock::Paragraph(TextContent::plain(
                "A major technology company today announced a significant breakthrough in artificial intelligence that could revolutionize healthcare, transportation, and manufacturing industries.".to_string()
            )),
            ContentBlock::Paragraph(TextContent::plain(
                "The new AI system demonstrates unprecedented capabilities in complex reasoning and problem-solving, potentially affecting millions of jobs worldwide.".to_string()
            )),
        ],
        metadata: ArticleMetadata {
            published_date: Some("2025-01-01T00:00:00.000000Z".to_string()),
            author: Some("Tech Reporter".to_string()),
            url: Some("https://techexample.com/ai-breakthrough".to_string()),
            feed_name: "Technology News".to_string(),
        },
        comments: vec![],
    };

    let political_article = Article {
        title: "International Trade Agreement Faces Opposition".to_string(),
        content: vec![
            ContentBlock::Paragraph(TextContent::plain(
                "A proposed international trade agreement has generated significant opposition from labor unions and environmental groups across three continents.".to_string()
            )),
        ],
        metadata: ArticleMetadata {
            published_date: Some("2025-01-01T00:00:00.000000Z".to_string()),
            author: Some("Political Reporter".to_string()),
            url: Some("https://newsexample.com/trade-agreement".to_string()),
            feed_name: "Political News".to_string(),
        },
        comments: vec![],
    };

    let health_article = Article {
        title: "Global Health Organization Issues New Guidelines".to_string(),
        content: vec![
            ContentBlock::Paragraph(TextContent::plain(
                "The World Health Organization has released updated vaccination guidelines that will impact public health policies in over 100 countries.".to_string()
            )),
        ],
        metadata: ArticleMetadata {
            published_date: Some("2025-01-01T00:00:00.000000Z".to_string()),
            author: Some("Health Reporter".to_string()),
            url: Some("https://healthexample.com/guidelines".to_string()),
            feed_name: "Health News".to_string(),
        },
        comments: vec![],
    };

    let tech_feed = Feed {
        name: "Technology News".to_string(),
        description: Some("Latest technology developments".to_string()),
        url: Some("https://techexample.com".to_string()),
        articles: vec![breaking_article],
    };

    let politics_feed = Feed {
        name: "Political News".to_string(),
        description: Some("Global political developments".to_string()),
        url: Some("https://newsexample.com".to_string()),
        articles: vec![political_article],
    };

    let health_feed = Feed {
        name: "Health News".to_string(),
        description: Some("Health and medical news".to_string()),
        url: Some("https://healthexample.com".to_string()),
        articles: vec![health_article],
    };

    Document {
        metadata: DocumentMetadata {
            title: "Daily News Digest".to_string(),
            author: "News Aggregator".to_string(),
            description: Some("Today's most important stories".to_string()),
            generated_at: "2025-01-01T00:00:00.000000Z".to_string(),
        },
        feeds: vec![tech_feed, politics_feed, health_feed],
    }
}

#[tokio::test]
async fn test_front_page_generation_with_ollama() {
    let provider = AiProvider::Ollama {
        base_url: "http://127.0.0.1:1234".to_string(),
        model: "llama2".to_string(),
    };

    let generator = FrontPageGenerator::new(provider).unwrap();
    let document = create_test_document();

    // This test requires a running Ollama server with temperature 0
    // Skip if server is not available to avoid CI failures
    if let Ok(front_page) = generator.generate_front_page(&document).await {
        // Normalize the output for consistent snapshots by removing dynamic elements
        let normalized_output = normalize_ai_output(&front_page);
        assert_snapshot!("front_page_generation_ollama", normalized_output);
    } else {
        // If Ollama server is not available, test the error case
        assert_snapshot!(
            "front_page_generation_ollama_unavailable",
            "Ollama server not available at http://127.0.0.1:1234"
        );
    }
}

#[test]
fn test_content_preparation() {
    let provider = AiProvider::Ollama {
        base_url: "http://127.0.0.1:1234".to_string(),
        model: "llama2".to_string(),
    };

    let generator = FrontPageGenerator::new(provider).unwrap();
    let document = create_test_document();

    let content = generator.prepare_content(&document).unwrap();

    // Normalize content by removing potential whitespace variations
    let normalized_content = normalize_markdown_content(&content);
    assert_snapshot!("content_preparation_markdown", normalized_content);
}

#[test]
fn test_prompt_construction() {
    let provider = AiProvider::Ollama {
        base_url: "http://127.0.0.1:1234".to_string(),
        model: "llama2".to_string(),
    };

    let generator = FrontPageGenerator::new(provider).unwrap();
    let test_content = "Sample news content for testing prompt construction";

    let prompt = generator.build_prompt(test_content);

    // Normalize the prompt for consistent snapshots
    let normalized_prompt = normalize_prompt_content(&prompt);
    assert_snapshot!("prompt_construction_template", normalized_prompt);
}

fn normalize_ai_output(output: &str) -> String {
    // Remove any timestamps, URLs, or other dynamic content that might vary between runs
    // This ensures consistent snapshot testing with AI-generated content
    output
        .lines()
        .map(|line| {
            // Remove specific URLs and replace with placeholders
            line.replace("https://techexample.com/ai-breakthrough", "[TECH_URL]")
                .replace("https://newsexample.com/trade-agreement", "[NEWS_URL]")
                .replace("https://healthexample.com/guidelines", "[HEALTH_URL]")
        })
        .collect::<Vec<_>>()
        .join("\n")
        .trim()
        .to_string()
}

fn normalize_markdown_content(content: &str) -> String {
    // Normalize markdown content for consistent snapshots
    content
        .lines()
        .map(|line| line.trim_end()) // Remove trailing whitespace
        .collect::<Vec<_>>()
        .join("\n")
        .trim()
        .to_string()
}

fn normalize_prompt_content(prompt: &str) -> String {
    // Normalize prompt content by replacing the dynamic test content with a placeholder
    prompt.replace(
        "Sample news content for testing prompt construction",
        "[TEST_CONTENT_PLACEHOLDER]",
    )
}

