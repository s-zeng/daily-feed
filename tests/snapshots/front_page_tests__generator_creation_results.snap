---
source: tests/front_page_tests.rs
expression: results
---
{
  "anthropic_success": true,
  "ollama_success": true
}
